node/core/dispute-coordinator/src/real/participation/mod.rs

```diff
@@ -0,0 +1,432 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashSet;
+
+use futures::{
+	channel::{mpsc, oneshot},
+	FutureExt, SinkExt,
+};
+
+use polkadot_node_primitives::{ValidationResult, APPROVAL_EXECUTION_TIMEOUT};
+use polkadot_node_subsystem::{
+	messages::{AvailabilityRecoveryMessage, AvailabilityStoreMessage, CandidateValidationMessage},
+	ActiveLeavesUpdate, RecoveryError, SubsystemContext, SubsystemSender,
+};
+use polkadot_node_subsystem_util::runtime::get_validation_code_by_hash;
+use polkadot_primitives::v1::{BlockNumber, CandidateHash, CandidateReceipt, Hash, SessionIndex};
+
+use crate::real::LOG_TARGET;
+
+use super::{
+	error::{Fatal, FatalResult, NonFatal, Result},
+	ordering::CandidateComparator,
+};
+
+mod queues;
+use queues::Queues;
+pub use queues::{Error as QueueError, ParticipationRequest};
+
+/// How many participation processes do we want to run in parallel the most.
+///
+/// This should be a relatively low value, while we might have a speedup once we fetched the data,
+/// due to multi core architectures, but the fetching itself can not be improved by parallel
+/// requests. This means that higher numbers make it harder for a single dispute to resolve fast.
+const MAX_PARALLEL_PARTICIPATIONS: usize = 3;
```

Comment:
Follow-up issue: make this configurable?

Reply:
Will file one.